custom/icds_reports/utils/aggregation_helpers/distributed/agg_dashboard_acitivity.py

```diff
@@ -0,0 +1,225 @@
+        return """
+        CREATE TABLE IF NOT EXISTS "{tablename}" (
+            CHECK (date = DATE %(date)s),
+            LIKE "{parent_tablename}" INCLUDING DEFAULTS INCLUDING CONSTRAINTS INCLUDING INDEXES
+        ) INHERITS ("{parent_tablename}")
+        """.format(
+            parent_tablename=self.aggregate_parent_table,
+            tablename=self.tablename,
+        ), {
+            "date": self.date.strftime("%Y-%m-%d"),
+        }
+
+    def get_last_agg_date(self):
+        from custom.icds_reports.models.aggregate import DashboardUserActivityReport
+        result = DashboardUserActivityReport.objects.filter(date__lt=self.date.strftime("%Y-%m-%d")).aggregate(Max('date'))
+        if result:
+            return result['date__max']
+        return None
+
+    def add_missing_users(self):
+        return """
+        INSERT INTO "{tablename}" (
+            username, state_id, district_id,block_id,
+            user_level,date
+        )
+        VALUES (%s,%s,%s,%s,%s, %s)
+        """.format(
+            tablename=self.tablename
+        ), {
+            'values': [tuple(list(loc) + [self.date]) for loc in self.get_user_locations()],
+        }
+
+    def rollover_previous_data(self):
+        query_param = {'date': self.date,
+                       'last_agg_date': self.last_agg_date}
+        return """
+        UPDATE "{tablename}" user_activity SET
+            location_launched = ut.location_launched,
+            last_activity = ut.last_activity
+        FROM  (
+        SELECT
+            username,
+            location_launched,
+            last_activity
+            FROM "{parent_tablename}" WHERE date = %(last_agg_date)s
+            )ut
+        WHERE user_activity.username = ut.username
+        """.format(
+            tablename=self.tablename,
+            parent_tablename=self.aggregate_parent_table
+        ), query_param
+
+    def update_query(self):
+        last_time_to_consider = self.date
+        latest_month = (last_time_to_consider - timedelta(days=1)).replace(day=1)
+
+        yield """
+        UPDATE "{tablename}" user_activity
+            SET location_launched = CASE WHEN num_launched_blocks>0 THEN TRUE ELSE FALSE END
+        FROM (
+            SELECT
+                block_id, district_id, state_id, num_launched_blocks,aggregation_level
+            FROM agg_awc where month=%(latest_month)s and aggregation_level<=3
+            ) ut
+        WHERE (
+            user_activity.user_level= ut.aggregation_level AND
+            user_activity.state_id = ut.state_id AND
+            user_activity.district_id = ut.district_id AND
+            user_activity.block_id = ut.block_id AND
+            user_activity.location_launched is not TRUE
+        )
+        """.format(
+            tablename=self.tablename
+        ), {
+            'latest_month': latest_month
+        }
+
+
+        # This is query I prepare which could do what the last two queries are doing but the cost of this single query
```

Comment:
i think it would be preferable to do one query over  2 if possible

Reply:
I did a sanity comparison, below two queries gave the cost of about 30Ks and this query of joining alone gave the cost of 30 lacs. which looks significantly higher.